src/daft-core/src/array/ops/utf8.rs

```diff
@@ -378,6 +414,162 @@ impl Utf8Array {
+            (self_len, pattern_len, replacement_len)
+                if self_len == pattern_len && self_len == replacement_len =>
+            {
+                if regex {
+                    let regex_iter = pattern_arrow.iter().map(|pat| pat.map(regex::Regex::new));
+                    regex_replace(self_arrow.iter(), regex_iter, replacement_arrow.iter(), self.name())
+                } else {
+                    replace_on_literal(self_arrow.iter(), pattern_arrow.iter(), replacement_arrow.iter(), self.name())
+
+                }
+            }
+            (1, pattern_len, replacement_len) if pattern_len == replacement_len => {
+                let self_scalar_value = self.get(0);
+                match self_scalar_value {
+                    None => Ok(Utf8Array::full_null(
+                        self.name(),
+                        self.data_type(),
+                        pattern_len,
+                    )),
+                    Some(self_v) => {
+                        let arr_iter = std::iter::repeat(Some(self_v)).take(pattern_len);
+                        if regex {
+                            let regexes = pattern_arrow.iter().map(|pat| pat.map(regex::Regex::new));
+                            regex_replace(arr_iter, regexes, replacement_arrow.iter(), self.name())
+                        } else {
+                            replace_on_literal(arr_iter, pattern_arrow.iter(), replacement_arrow.iter(), self.name())
+                        }
+                    }
+                }
+            }
+            (self_len, 1, replacement_len) if self_len == replacement_len => {
+                let pattern_scalar_value = pattern.get(0);
+                match pattern_scalar_value {
+                    None => Ok(Utf8Array::full_null(
+                        self.name(),
+                        self.data_type(),
+                        self_len,
+                    )),
+                    Some(pattern_v) => {
+                        if regex {
+                            let re = Some(regex::Regex::new(pattern_v));
+                            let regex_iter = std::iter::repeat(re).take(self_len);
+                            regex_replace(self_arrow.iter(), regex_iter, replacement_arrow.iter(), self.name())
+                        } else {
+                            let pattern_iter = std::iter::repeat(Some(pattern_v)).take(self_len);
+                            replace_on_literal(self_arrow.iter(), pattern_iter, replacement_arrow.iter(), self.name())
+                        }
+                    }
+                }
+            }
+            (self_len, pattern_len, 1) if self_len == pattern_len => {
+                let replacement_scalar_value = replacement.get(0);
+                match replacement_scalar_value {
+                    None => Ok(Utf8Array::full_null(
+                        self.name(),
+                        self.data_type(),
```

Comment:
This is dangerous, we probably want to explicitly use the String type here?

Ditto for the other locations where we use `Utf8Array::full_null`